plugins/hls-semantic-tokens-plugin/test/Main.hs

```diff
@@ -141,14 +121,49 @@ semanticTokensValuePatternTests =
       goldenWithSemanticTokens "pattern bind" "TPatternbind"
     ]
 
+mkSemanticConfig :: Value -> Config
+mkSemanticConfig setting = def{plugins = Map.insert "SemanticTokens" conf (plugins def)}
+    where
+      conf = def{plcConfig = (\(Object obj) -> obj) setting }
+
+modifySemantic :: Value -> Session ()
+modifySemantic setting = Test.setHlsConfig $ mkSemanticConfig setting
+
+
+directFile :: FilePath -> Text -> [FS.FileTree]
+directFile fp content =
+  [ FS.directCradle [Text.pack fp]
+  , file fp (text content)
+  ]
+
+semanticTokensConfigTest :: TestTree
+semanticTokensConfigTest = testGroup "semantic token config test" [
+        testCase "function to variable" $ do
+            let content = Text.unlines ["module Hello where", "go _ = 1"]
+            let fs = mkFs $ directFile "Hello.hs" content
+            let funcVar = object [ "tokenMapping" .= object ["function" .= var] ]
+                var :: String
+                var = "variable"
+            do
+                recorder <- pluginTestRecorder
+                Test.Hls.runSessionWithServerInTmpDir' (semanticTokensPlugin recorder)
+                    (mkSemanticConfig funcVar)
+                    def {ignoreConfigurationRequests = False}
+                    fullCaps
+                    fs $ do
+                    -- modifySemantic funcVar
+                    void waitForBuildQueue
+                    doc <- openDoc "Hello.hs" "haskell"
+                    void waitForBuildQueue
+                    result1 <- docLspSemanticTokensString doc
+                    liftIO $ unlines (map show result1) @?= "2:1-3 SemanticTokenTypes_Variable \"go\"\n"
```

Comment:
so here for example if you did the backwards mapping you'd presumably get the wrong thing?

Reply:
Yes, we should not do backward after customize the config. Maybe we should make a note somewhere.